cpp/src/shuffler/postbox.cpp

```diff
@@ -12,34 +12,100 @@
 namespace rapidsmpf::shuffler::detail {
 
 
+PostBox::PostBox(Rank nranks, std::function<Rank(PartID)>&& partition_owner)
+    : partition_owner_{std::move(partition_owner)} {
+    pigeonhole_.reserve(static_cast<std::size_t>(nranks));
+}
+
 void PostBox::insert(Chunk&& chunk) {
     std::lock_guard const lock(mutex_);
-    auto [_, inserted] = pigeonhole_[chunk.pid].insert({chunk.cid, std::move(chunk)});
-    RAPIDSMPF_EXPECTS(inserted, "PostBox.insert(): chunk already exist");
+    Rank owner = partition_owner_(chunk.pid);
+    auto [_, inserted] = chunk_ids_.emplace(chunk.cid);
+    RAPIDSMPF_EXPECTS(
+        inserted, "PostBox.insert(): chunk already exist " + std::to_string(chunk.cid)
+    );
+
+    // insert into the rank-based pigeonhole
+    pigeonhole_[owner].emplace_back(std::move(chunk));
 }
 
 Chunk PostBox::extract(PartID pid, ChunkID cid) {
```

Comment:
Again, this function is only called by the `postbox_spilling` so if `PostBox::search` were adapted to return rank, id pairs, we wouldn't need to change this definition really.